Moving telemetry to MSAL

trwalke opened the pull request:
Removing Telemetry from core and moving it to MSAL. Core will now use CoreTelemetry which will be injected with either the ADAL or MSAL telemetry implementation on startup. This way, MSAL will have a public telemetry class (MSALTelemetry) and core will have access to only internal telemetry classes (CoreTelemetry, iTelemetry)

VSTS 510046

Commit:
Moving telemetry to MSAL

henrik-me commented on core/src/OAuth2/OAuth2Client.cs:
```diff
@@ -92,8 +88,8 @@ internal async Task<T> ExecuteRequestAsync<T>(Uri endPoint, HttpMethod method, R
-            var httpEvent = new HttpEvent() { HttpPath = endpointUri, QueryParams = endpointUri.Query };
-            Client.Telemetry.GetInstance().StartEvent(requestContext.TelemetryRequestId, httpEvent);
+            //var httpEvent = new HttpEvent() { HttpPath = endpointUri, QueryParams = endpointUri.Query };
+            //Client.Telemetry.GetInstance().StartEvent(requestContext.TelemetryRequestId, httpEvent);
```
? #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:
```diff
@@ -59,6 +59,14 @@
+    <None Remove="Telemetry\Event.cs" />
+    <None Remove="Telemetry\HttpEvent.cs" />
+    <None Remove="Telemetry\TelemetryTokenCacheAccessor.cs" />
+    <None Remove="Telemetry\UiEvent.cs" />
```
? .. Telemetry/*.cs #Closed

trwalke replied:
Removed
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/Telemetry.cs:
```diff
@@ -25,11 +25,11 @@
 //
 //------------------------------------------------------------------------------
 
+using Microsoft.Identity.Client;
```
is this needed? #Closed

trwalke replied:
No
#Resolved #Closed

henrik-me reviewed (changes requested):
:clock1:

Commit:
Added implementation injection for core telemetry.

Commit:
Adding test implementation for CoreTelemetry

bgavrilMS commented on core/src/CoreTelemetry.cs:
```diff
@@ -0,0 +1,13 @@
+{
+    internal class CoreTelemetry
+    {
+        public static iTelemetry Instance;
```
please use ITelemetry #Closed

trwalke replied:
#Resolved #Closed

bgavrilMS commented on core/src/Microsoft.Identity.Core.csproj:
```diff
@@ -127,4 +127,6 @@
       <Version>27.0.2</Version>
     </PackageReference>
   </ItemGroup>
+	  <Target Name="ProjectReferenceSolution" BeforeTargets="ResolveAssemblyReferences" DependsOnTargets="ResolveProjectReferences">
```
What is the purpose of this? #Closed

trwalke replied:
This solves one of the visual studio bugs that results in broken project references in multi-targeted solutions. I only put it into test projects as they are the only ones effected. #Closed

trwalke replied:
#Resolved #Closed

bgavrilMS commented on core/tests/Test.Microsoft.Identity.Core.Unit/Test.Microsoft.Identity.Core.Unit.csproj:
```diff
@@ -83,4 +83,6 @@
   <ItemGroup>
     <Service Include="{82a7f48d-3b50-4b1e-b82e-3ada8210c358}" />
   </ItemGroup>
+  <Target Name="ProjectReferenceSolution" BeforeTargets="ResolveAssemblyReferences" DependsOnTargets="ResolveProjectReferences">
```
same as above #Closed

trwalke replied:
#Resolved #Closed

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Microsoft.Identity.Client.csproj:
```diff
@@ -59,6 +59,14 @@
     <None Include="**\*.cs;**\*.xml;**\*.axml" Exclude="obj\**\*.*;bin\**\*.*" />
     <Compile Remove="Platforms\**\*.*" />
     <Compile Remove="Features\**\*.*" />
+    <None Remove="Telemetry.cs" />
```
You probably do not need these at all. #Closed

trwalke replied:
Auto generated by visual studio. I can remove #Closed

bgavrilMS commented on msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:
```diff
@@ -204,6 +204,21 @@ private List<EventBase> CollateOrphanedEvents(String requestId)
+            return GetInstance();
+        }
+
+        void iTelemetry.StartEvent(string requestId, EventBase eventToStart)
```
There is no need to explicitly implement interfaces unless your object implements multiple interfaces and 2 of those interfaces define a method that has the same name. 

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/interfaces/explicit-interface-implementation #Closed

trwalke replied:
This is needed because the itelemetry interface is internal only as we are trying to remove public classes from core. by implementing the interface explicitly,  it allows itelemetry to be internal and used by core while MSALTelemetry can remain public and not be exposed to core. #Closed

bgavrilMS replied:
Ah, sneaky... :)
 #Closed

henrik-me replied:
Why can't the Implicit implementation give the same (not exposed outside the library)? An interface can be internally implemented

---
In reply to: [208693813](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1173#discussion_r208693813) [](ancestors = 208693813)

bgavrilMS commented on core/src/CoreTelemetry.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Microsoft.Identity.Core
+{
+    internal class CoreTelemetry
```
I had a look at this and have a few suggestions: 

1. the constructor of CoreTelemetry should take in an ITelemetry param. This way your class is immutable and easier to use. You should also expose properties not members, i.e. public static ITelemetry Instance { get; private set;}
2. I don't like that you have to do CoreTelemetry.Instance.GetInstance() ... Ideally we'd just have CoreTelemetry.Instance
3. The interface ITelemetry should not need to have a GetInstance()
4. It would be cleaner if the classes MSALTelemetry and ADALTelemetry would not be singletons but normal classes with state etc. and you'd use CoreTelemetry everywhere. Not sure how much work this would be though, but if it's not a lot, it's probably worth it. #Closed

trwalke replied:
1. CoreTelemetry.instance is static and no object reference is available for coreTelemetry. I propose something like public static void InitializeCoreTelemetryService(ITelemetry instance)?
2. #Resolved
3. #Resolved
4. This looks like it can get tricky, i think I will keep it as is now for the sake of time and create a task tracking this. #Closed

bgavrilMS commented on msal/src/Microsoft.Identity.Client/Internal/ModuleInitializer.cs:
```diff
@@ -56,6 +56,7 @@ public static void EnsureModuleInitialized()
                 if (!isInitialized)
                 {
                     CoreExceptionFactory.Instance = new MsalExceptionFactory();
+                    CoreTelemetry.Instance = new MSALTelemetry();
```
Hmm, this is possibly a bug - you are creating a new instance of MSALTelemetry and also a new instance in the MSALTelemetry class. So in fact you have 2 singletons - one in MSAL and one in core for the same object. 

I see 2 ways out of this: 

1. As stated above, refactor MSAL/ADAL Telemetry objects to not be singletons
2. or initialize CoreTelemetry.Instance with MsalTelemetry.Singleton #Closed

trwalke replied:
#resolved with option 2 #Closed

bgavrilMS commented on core/src/CoreTelemetry.cs:
```diff
@@ -0,0 +1,13 @@
+
+namespace Microsoft.Identity.Core
+{
+    internal class CoreTelemetry
```
Hmm, how about CoreTelemetryService .... this name makes it a bit clearer that it is a singleton... ? #Closed

trwalke replied:
I Agree #Closed

Commit:
Merge remote-tracking branch 'origin/dev' into trwalke/MoveTelemetryToMSAL

# Conflicts:
#	msal/src/Microsoft.Identity.Client/Internal/ModuleInitializer.cs
#	msal/src/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

Commit:
Adding initializer to CoreTelemetryService.
Changing CoreTelemetry to CoreTelemetryService.
Addressing PR Comments.
Removing GetInstance from ITelemetry
renaming iTelemetry to ITelemetry.

henrik-me commented on core/src/CoreTelemetryService.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace Microsoft.Identity.Core
+{
+    internal class CoreTelemetryService
```
mark class as static #Closed

trwalke replied:
#Resolved #Closed

henrik-me replied:
not what I meant. Everywhere else we have a mathod called GetInstance. Personally I prefer the property called Instance. Most of our similar code follows the pattern of a method named GetInstance() ... thus property = Intance, or method = GetInstance()

---
In reply to: [208828711](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1173#discussion_r208828711) [](ancestors = 208828711)

henrik-me commented on core/src/Telemetry/iTelemetry.cs:
```diff
@@ -0,0 +1,15 @@
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Core
```
Filename seems still to be with small i #Closed

trwalke replied:
#Resolved #Closed

henrik-me replied:
Seems still not updated. But this might be a tools issue more than anything.

---
In reply to: [208828746](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1173#discussion_r208828746) [](ancestors = 208828746)

henrik-me commented:
> using System;

missing file headers #Closed

---
Refers to: core/src/Telemetry/iTelemetry.cs:1 in 1c375ba. [](commit_id = 1c375ba7d7d72aed3b90bdf58a85ea521ad38304, deletion_comment = False)

henrik-me commented on core/src/Telemetry/CoreTestTelemetryService.cs:
```diff
@@ -0,0 +1,16 @@
+
+namespace Microsoft.Identity.Core
+{
+    internal class CoreTestTelemetryService : ITelemetry
```
>TestT [](start = 23, length = 5)

Why have a test telemetry service here? #Closed

trwalke replied:
No longer needed 
#Resolved #Closed

henrik-me commented:
All new files are missing file headers #Closed

henrik-me commented:
Why are you moving all the telemetry files out of the telemetry namespace? #Closed

henrik-me commented on core/src/CoreTelemetryService.cs:
```diff
@@ -0,0 +1,18 @@
+    {
+        public static void InitializeCoreTelemetryService(ITelemetry instance)
+        {
+            Instance = instance;
```
>instance [](start = 23, length = 8)

pls. ensure this is initialized in the ModuleInitalizer of ADAL as well, as otherwise parts of the code will likely crash because the instance is not set. 
Note: only an issue if the Telemetry class is called in Core... which I assume it is given we have this service.

Aside: in the other classes the Instance is a method and called GetInstance, perhaps we should do the same here for consistency? #Closed

trwalke replied:
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:
```diff
@@ -55,15 +55,15 @@ public void RegisterReceiver(Receiver r)
         }
 
-        private static readonly Telemetry Singleton = new Telemetry();
+        private static readonly MSALTelemetry Singleton = new MSALTelemetry();
```
>MSALTelemetry [](start = 32, length = 13)

ITelemetry ?  #Closed

trwalke replied:
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:
```diff
@@ -25,18 +25,18 @@
     /// 
     /// </summary>
-    public class Telemetry
+    internal class MSALTelemetry : ITelemetry
```
>MSALTelemetry [](start = 19, length = 13)

Not sure this should be renamed. that will be another breaking change. The namespace takes care of separation #Closed

trwalke replied:
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:
```diff
@@ -25,18 +25,18 @@
     /// 
     /// </summary>
-    public class Telemetry
+    internal class MSALTelemetry : ITelemetry
```
>internal [](start = 4, length = 8)

I suppose it has to be public here. How are you exposing this? #Closed

trwalke replied:
Yes, exposing here 
#Resolved #Closed

henrik-me commented:
>         public bool TelemetryOnFailureOnly { get; set; }

Any reason you are not moving these to Core? #Closed

---
Refers to: core/src/Telemetry.cs:74 in 1c375ba. [](commit_id = 1c375ba7d7d72aed3b90bdf58a85ea521ad38304, deletion_comment = False)

henrik-me commented on msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:
```diff
@@ -55,15 +55,15 @@ public void RegisterReceiver(Receiver r)
         }
 
-        private static readonly Telemetry Singleton = new Telemetry();
+        private static readonly MSALTelemetry Singleton = new MSALTelemetry();
```
>Singleton [](start = 46, length = 9)

personal preference, call this instance #Closed

trwalke replied:
I agree
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/Internal/ModuleInitializer.cs:
```diff
@@ -56,6 +56,7 @@ public static void EnsureModuleInitialized()
                 if (!isInitialized)
                 {
                     CoreExceptionFactory.Instance = new MsalExceptionFactory();
+                    CoreTelemetryService.InitializeCoreTelemetryService(MSALTelemetry.GetInstance());
```
>InitializeCoreTelemetryService(MSALTelemetry.GetInstance()); [](start = 41, length = 60)

Can we do like for the exception factory?

henrik-me commented:
imo, not ready yet #Closed

trwalke commented:
"public bool TelemetryOnFailureOnly { get; set; }
Any reason you are not moving these to Core?"
Refers to: msal/src/Microsoft.Identity.Client/MSALTelemetry.cs:74


@henrik-me Do you mean defining these in ITelemetry so that core has access to them? at the moment, only StartEvent() and EndEvent() are used by core.  #Closed

trwalke commented:
> Why are you moving all the telemetry files out of the telemetry namespace?
@henrik-me Refactoring error, resolving #Closed

Commit:
Changing core telemetry classes to Microsoft.Identity.Client.Telemetry

Changing core telemetry to static.

Updating class/method names

Commit:
Removing unnecessary file

trwalke commented:
Added missing file headers #Closed

Commit:
Merge remote-tracking branch 'origin/dev' into trwalke/MoveTelemetryToMSAL

# Conflicts:
#	core/tests/Test.Microsoft.Identity.Core.Unit/InstanceTests/AadAuthorityTests.cs
#	msal/tests/Test.MSAL.NET.Unit/Test.MSAL.NET.Unit.csproj

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Telemetry.cs:
```diff
@@ -0,0 +1,56 @@
+
+namespace Microsoft.IdentityModel.Clients.ActiveDirectory
+{
+    internal class Telemetry : ITelemetry
```
>internal class Telemetry : ITelemetry [](start = 4, length = 37)

pls. add comments to the class. and espcially a remarks section explaining why the implementation is blank #Closed

trwalke replied:
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/Telemetry.cs:
```diff
@@ -65,7 +65,7 @@ public void RegisterReceiver(Receiver r)
         /// <returns></returns>
         public static Telemetry GetInstance()
```
>GetInstance() [](start = 32, length = 13)

implemented as a method... then it's called GetInstance. This is the most used pattern in the codebase. pls. stick to this one #Closed

trwalke replied:
#Resolved #Closed

henrik-me commented on msal/src/Microsoft.Identity.Client/Telemetry.cs:
```diff
@@ -204,6 +204,16 @@ private List<EventBase> CollateOrphanedEvents(String requestId)
+        void ITelemetry.StopEvent(string requestId, EventBase eventToStop)
+        {
+            StopEvent(requestId, eventToStop);
+        }
```
still I do not get why this is needed. it's already implemented implicitly #Closed

trwalke replied:
See other comment
 #Closed

henrik-me commented:
>             Assert.AreEqual(logger, CoreLoggerBase.Default, "Initialization should have happened only once");

seems like you are not asserting that Telemetry is initialized

#Resolved
---
Refers to: msal/tests/Test.MSAL.NET.Unit/ModuleInitializerTest.cs:60 in 549d264. [](commit_id = 549d264798e512526bb169cf5b120aa21cf865dd, deletion_comment = False) #Closed

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Telemetry.cs:
```diff
@@ -0,0 +1,56 @@
+        public static Telemetry GetInstance()
+        {
+            return Instance as Telemetry;
+        }
```
:) #Closed

henrik-me commented on adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Telemetry.cs:
```diff
@@ -0,0 +1,56 @@
+            return Instance as Telemetry;
+        }
+
+        void ITelemetry.StartEvent(string requestId, EventBase eventToStart)
```
>ITelemetry [](start = 13, length = 10)

why explicitly? #Closed

trwalke replied:
@henrik-me 
With the intention of making Telemetry.cs internal, I did not want to publicly expose the new ITelemetry interface to ADAL. Thus, the interface is internal. However, public Telemetry.cs implements the internal ITelemetry interface; the only way that I have found to do that is for the Telemetry to implement the interface explicitly. 
This way, 
1. Customers using MSAL can use the public Telemetry.cs, 
2. Core can still use the iTelemetry interface and have the implementation injected
3. Customers using ADAL will not have access to Telemetry.CS or ITelemetry.cs as neither of the two classes are used in any meaningful way in ADAL.

We can get around the explicit implementation by making ITelemetry public but we will then expose it to ADAL. 
Thoughts? #Closed

henrik-me replied:
the class is internal, it will not be public. This should not have anything to do with the interface being implicitly implemented. I don't get why it has to be public to be explicitly implemented

---
In reply to: [209065629](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1173#discussion_r209065629) [](ancestors = 209065629)

trwalke replied:
@henrik-me Ah, I understand, I confused this with the other public Telemetry class in MSAL, you are right, it does not need to be implemented explicitly. I will remove #Closed

trwalke replied:
#Resolved #Closed

trwalke committed:
Merge branch 'dev' into trwalke/MoveTelemetryToMSAL

Commit:
Changing getinstance to method.

Adding assertion for CoreTelemetry in moduleInitializerTests.

Adding comments to ADAL Telemetry class.

henrik-me commented on core/src/CoreTelemetryService.cs:
```diff
@@ -0,0 +1,51 @@
+
+namespace Microsoft.Identity.Core
+{
+    internal static class CoreTelemetryService
```
>static [](start = 13, length = 6)

... I think you are running into the issues because the class is static... probably that should be removed. #Closed

trwalke replied:
#Resolved #Closed

Commit:
Changing ADAL Itelemetry implementation to implicit.

Removing static from CoreTelemetryService.

henrik-me reviewed (approved):
:shipit:
